### PR TITLE
[WIP] FWaaSv2.0 Policy CRUD

### PR DESCRIPTION
**Pull request by Elethiomel**
For #514 

https://github.com/openstack/neutron-lib/blob/56033ba643812a30577f6ab17648806c2ee494ba/neutron_lib/api/definitions/firewall_v2.py#L159
https://github.com/openstack/neutron-fwaas/blob/27906d0acff349c9e5ac7955d203fb2233c295db/neutron_fwaas/db/firewall/v2/firewall_db_v2.py#L148

### CODE REVIEW

**Comment by Elethiomel** [comment]
@jtopjian Here's the entire Policy CRUD one for you :)

File: openstack/networking/v2/extensions/fwaas_v2/policies/results.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+// IsEmpty checks whether a PolicyPage struct is empty.
+func (r PolicyPage) IsEmpty() (bool, error) {
+	is, err := ExtractPolicies(r)
+	return len(is) == 0, err
+}
+
+// ExtractPolicies accepts a Page struct, specifically a RouterPage struct,
+// and extracts the elements into a slice of Router structs. In other words,
+// a generic collection is mapped into a relevant slice.
```
[the remark this answers is not in the document]
Comment (Elethiomel) [reply openstack/networking/v2/extensions/fwaas_v2/policies/results.go]: Resolved in commit 75ea88c2d2fc10b19e5f314af011d9223acdfcf0

File: openstack/networking/v2/extensions/fwaas_v2/policies/requests.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+	ID           string `json:"firewall_rule_id" required:"true"`
+	BeforeRuleID string `json:"insert_before,omitempty"`
+	AfterRuleID  string `json:"insert_after,omitempty"`
+}
+
+func (opts InsertRuleOpts) ToFirewallPolicyInsertRuleMap() (map[string]interface{}, error) {
+	return gophercloud.BuildRequestBody(opts, "")
+}
+
+func AddRule(c *gophercloud.ServiceClient, id string, opts InsertRuleOptsBuilder) (r InsertRuleResult) {
```
[the remark this answers is not in the document]
Comment (Elethiomel) [reply openstack/networking/v2/extensions/fwaas_v2/policies/requests.go]: Resolved in commit b73ae0590095b627087c5335777024c86d1aebb3

File: openstack/networking/v2/extensions/fwaas_v2/policies/requests.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+		r.Err = err
+		return
+	}
+	_, r.Err = c.Put(insertURL(c, id), b, &r.Body, &gophercloud.RequestOpts{
+		OkCodes: []int{200},
+	})
+	return
+}
+
+func RemoveRule(c *gophercloud.ServiceClient, id, ruleID string) (r RemoveRuleResult) {
```
[the remark this answers is not in the document]
Comment (Elethiomel) [reply openstack/networking/v2/extensions/fwaas_v2/policies/requests.go]: Resolved in commit 5f3e0f7091ecb2a2f32ae69a8726c35c2103e565

**Comment by chrischdi** [comment]
Hi @Elethiomel are you still into this?

**Comment by Elethiomel** [comment]
Yes I hope to progress on it soon. I've had some other major projects in the way unfortunately

File: openstack/networking/v2/extensions/fwaas_v2/policies/requests.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	ToFirewallPolicyUpdateMap() (map[string]interface{}, error)
+}
+
+// UpdateOpts contains the values used when updating a firewall policy.
+type UpdateOpts struct {
+	Name        *string  `json:"name,omitempty"`
+	Description *string  `json:"description,omitempty"`
+	Shared      *bool    `json:"shared,omitempty"`
+	Audited     *bool    `json:"audited,omitempty"`
+	Rules       []string `json:"firewall_rules,omitempty"`
```
[the remark this answers is not in the document]
Comment (codemanufaktur) [reply openstack/networking/v2/extensions/fwaas_v2/policies/requests.go:114]: fixed with https://github.com/Daimler/gophercloud/commit/7efe16ba8ad4061a39fa4c92bcd69b8a23b58ba8, but the acceptance test is probably missing. will check this on  #2057

**Comment by codemanufaktur** [comment]
@Elethiomel  #2057 is merged, I think you can close this one. Thank you :-)

**Comment by jtopjian** [comment]
Thank you both for your work on this.

**Pull request closed without merging**